src/db_cache/serde.rs

```diff
@@ -0,0 +1,283 @@
+//! Serialize/Deserialize impls for CachedKey and CachedEntry to support DbCache impls that
+//! can store cache items on-disk. Internally, the Serialize/Deserialize impls work by
+//! converting these types to representations that derive Serialize/Deserialize. The purpose
+//! of the indirection is to decouple the serialized format from the in-memory representation
+//! used by the rest of the codebase.
+
+use crate::block::Block;
+use crate::db_cache::{CachedEntry, CachedItem, CachedKey};
+use crate::db_state::SsTableId;
+use crate::filter::BloomFilter;
+use crate::flatbuffer_types::SsTableIndexOwned;
+use bytes::Bytes;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::sync::Arc;
+use ulid::Ulid;
+
+#[derive(Serialize, Deserialize)]
+enum SerializedSsTableId {
+    Wal(u64),
+    Compacted(Ulid),
+}
+
+impl From<SerializedSsTableId> for SsTableId {
+    fn from(value: SerializedSsTableId) -> Self {
+        match value {
+            SerializedSsTableId::Wal(id) => SsTableId::Wal(id),
+            SerializedSsTableId::Compacted(id) => SsTableId::Compacted(id),
+        }
+    }
+}
+
+impl From<SsTableId> for SerializedSsTableId {
+    fn from(value: SsTableId) -> Self {
+        match value {
+            SsTableId::Wal(id) => SerializedSsTableId::Wal(id),
+            SsTableId::Compacted(id) => SerializedSsTableId::Compacted(id),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+enum SerializedCachedKey {
+    V1(SerializedSsTableId, u64 /*#[default] String*/),
+}
+
+impl From<SerializedCachedKey> for CachedKey {
+    fn from(value: SerializedCachedKey) -> Self {
+        match value {
+            SerializedCachedKey::V1(sst_id, block_id) => CachedKey(sst_id.into(), block_id),
+        }
+    }
+}
+
+impl From<CachedKey> for SerializedCachedKey {
+    fn from(value: CachedKey) -> Self {
+        SerializedCachedKey::V1(value.0.into(), value.1)
+    }
+}
+
+impl Serialize for CachedKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let serialized_key: SerializedCachedKey = self.clone().into();
+        serialized_key.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for CachedKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let serialized_key = SerializedCachedKey::deserialize(deserializer)?;
+        Ok(serialized_key.into())
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+enum SerializedCachedEntryV1 {
+    Block(Bytes),
+    SsTableIndex(Bytes),
+    BloomFilter(Bytes),
+}
+
+impl SerializedCachedEntryV1 {
+    fn into_cached_entry(self) -> Result<CachedEntry, String> {
```

Comment:
Seems a little weird to return a String for error type here. Why not map to a SlateDBError?

Reply:
just noticed this, using a `DbCacheError` seems having less surprise on this 😲

Reply:
So the error returned here just gets passed to the serde Error impl's `custom` constructor ( https://docs.rs/serde/1.0.219/serde/de/trait.Error.html#tymethod.custom) which just expects an `impl Display` which then just gets wrapped in the serde implementation's error type. The implementations all seem to just call `to_string` on the value thats passed in and then pass that to the serde's error type. That said, there does seem to be a requirement that the error be formatted as lowercase and not ending in a `.` so I'll make sure that's the case.